### ISSUE TITLE
TSLint only Report Undefined Identifier Error

### ISSUE THREAD

**Issue by yudisupriyadi123** [w0rp/ale#928]
The typescript linter only gives me errors when identifier is undefined. Missing semicolon is not reported, curly rules and others:

```
# tslint.json
{
  "rulesDirectory": [
    "node_modules/codelyzer"
  ],
  "rules": {
    "curly": true,
    ....
``` 

:ALEInfo output
```
 Current Filetype: typescript
Available Linters: ['eslint', 'tslint', 'tsserver', 'typecheck']
  Enabled Linters: ['eslint', 'tslint', 'tsserver', 'typecheck']
 Linter Variables:
let g:ale_typescript_tslint_config_path = ''
let g:ale_typescript_tslint_executable = 'tslint'
let g:ale_typescript_tslint_ignore_empty_files = 0
let g:ale_typescript_tslint_rules_dir = ''
let g:ale_typescript_tslint_use_global = 0
let g:ale_typescript_tsserver_config_path = ''
let g:ale_typescript_tsserver_executable = 'tsserver'
let g:ale_typescript_tsserver_use_global = 0
 Global Variables:
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%s'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {}
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_linter_aliases = {}
let g:ale_linters = {}
let g:ale_open_list = 1
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0                                                                                                                                                
let g:ale_set_signs = 1                                                                                                                                                   
let g:ale_sign_column_always = 0                                                                                                                                          
let g:ale_sign_error = '>>'                                                                                                                                               
let g:ale_sign_offset = 1000000                                                                                                                                           
let g:ale_sign_warning = '--'                                                                                                                                             
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']                                                                                                      
let g:ale_warn_about_trailing_whitespace = 1                                                                                                                              
  Command History:                                                                                                                                                        
(started) ['/usr/bin/zsh', '-c', 'cd ''/home/yudi099/diajar/angular/base-chat/src/app'' && /home/yudi099/diajar/angular/base-chat/node_modules/.bin/tslint --format json -c ''/home/yudi099/diajar/angular/base-chat/tslint.json'' ''/tmp/vtGUUJx/180/app.component.ts''']                                                                          
(executable check - failure) typecheck                                                                                                                                    
(executable check - failure) eslint                                                                                                                                       
(finished - exit code 2) ['/usr/bin/zsh', '-c', 'cd ''/home/yudi099/diajar/angular/base-chat/src/app'' && /home/yudi099/diajar/angular/base-chat/node_modules/.bin/tslint 
--format json -c ''/home/yudi099/diajar/angular/base-chat/tslint.json'' ''/tmp/vtGUUJx/181/app.component.ts''']                                                           
<<<NO OUTPUT RETURNED>>>                                                                                                                                                  
(executable check - failure) typecheck                                                                                                                                    
(executable check - failure) eslint                                                                                                                                       
(started) ['/usr/bin/zsh', '-c', 'cd ''/home/yudi099/diajar/angular/base-chat/src/app'' && /home/yudi099/diajar/angular/base-chat/node_modules/.bin/tslint --format json -
c ''/home/yudi099/diajar/angular/base-chat/tslint.json'' ''/tmp/vtGUUJx/182/app.component.ts''']                                                                          
(executable check - failure) typecheck                                                                                                                                    
(executable check - failure) eslint                                                                                                                                       
(started) ['/usr/bin/zsh', '-c', 'cd ''/home/yudi099/diajar/angular/base-chat/src/app'' && /home/yudi099/diajar/angular/base-chat/node_modules/.bin/tslint --format json -
c ''/home/yudi099/diajar/angular/base-chat/tslint.json'' ''/tmp/vtGUUJx/183/app.component.ts''']                                                                          
(executable check - failure) typecheck                                                                                                                                    
(executable check - failure) eslint                                                                                                                                       
(finished - exit code 2) ['/usr/bin/zsh', '-c', 'cd ''/home/yudi099/diajar/angular/base-chat/src/app'' && /home/yudi099/diajar/angular/base-chat/node_modules/.bin/tslint 
--format json -c ''/home/yudi099/diajar/angular/base-chat/tslint.json'' ''/tmp/vtGUUJx/184/app.component.ts''']                                                           
<<<NO OUTPUT RETURNED>>>                                                                                                                                                  
(executable check - failure) typecheck                                                                                                                                    
(executable check - failure) eslint                                                                                                                                       
(finished - exit code 2) ['/usr/bin/zsh', '-c', 'cd ''/home/yudi099/diajar/angular/base-chat/src/app'' && /home/yudi099/diajar/angular/base-chat/node_modules/.bin/tslint 
--format json -c ''/home/yudi099/diajar/angular/base-chat/tslint.json'' ''/tmp/vtGUUJx/185/app.component.ts''']                                                           
<<<NO OUTPUT RETURNED>>>                                                                                                                                                  
(executable check - failure) typecheck                                                                                                                                    
(executable check - failure) eslint                                                                                                                                       
(finished - exit code 2) ['/usr/bin/zsh', '-c', 'cd ''/home/yudi099/diajar/angular/base-chat/src/app'' && /home/yudi099/diajar/angular/base-chat/node_modules/.bin/tslint 
--format json -c ''/home/yudi099/diajar/angular/base-chat/tslint.json'' ''/tmp/vtGUUJx/186/app.component.ts''']                                                           
<<<NO OUTPUT RETURNED>>>                                                                                                                                                  
(executable check - failure) typecheck             
```

Running the linter directly gives me what I expect (missing semicolon):
```
$ /home/yudi099/diajar/angular/base-chat/node_modules/.bin/tslint --format json -c ../../tslint.json app.component.ts
Warning: The 'no-use-before-declare' rule requires type checking
[{"endPosition":{"character":26,"line":15,"position":306},"failure":"Missing semicolon","fix":{"innerStart":306,"innerLength":0,"innerText":";"},"name":"app.component.ts","ruleName":"semicolon","ruleSeverity":"ERROR","startPosition":{"character":26,"line":15,"position":306}}]
```

**Comment by w0rp** [comment]
Could you create a project which replicates that error, and share that? I can't see what the problem is here.

**Comment by yudisupriyadi123** [comment]
Should I upload my fake project to github?

**Comment by w0rp** [comment]
Sure, that'll work.

**Comment by yudisupriyadi123** [comment]
this is it: https://github.com/yudisupriyadi123/fake-angular
I found that not only undefined, another error also reported but some rules not.

**Comment by w0rp** [comment]
Could you add package.json or something too? I don't know what needs to be installed here.

**Comment by yudisupriyadi123** [comment]
I open /tmp/fake-angular/app.component.ts and errors showed without installing anything.

**Comment by w0rp** [comment]
Your `tsconfig.json` file uses a rules directory from `node_modules`. I'll need at least a `package.json` file for your project so I can install everything you installed to `node_modules`.

**Comment by yudisupriyadi123** [comment]
New repo: https://github.com/yudisupriyadi123/fake-angular-v4
In `src/app/app.component.ts` I have added some test.

**Comment by ryanbas21** [comment]
Not sure if this is related but I am getting `unknown function: tslint` 

`ALEInfo
Current Filetype: typescript
Available Linters: ['eslint', 'tslint', 'tsserver', 'typecheck', 'eslint', 'tslint', 'tsserver', 'typecheck']
  Enabled Linters: ['tslint']
 Linter Variables:
let g:ale_typescript_tslint_config_path = ''
let g:ale_typescript_tslint_executable = 'tslint'
let g:ale_typescript_tslint_ignore_empty_files = 0
let g:ale_typescript_tslint_rules_dir = ''
let g:ale_typescript_tslint_use_global = 0
 Global Variables:
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%s'
let g:ale_echo_msg_warning_str = 'W'
let g:ale_enabled = 1
let g:ale_fix_on_save = 1
let g:ale_fixers = {'typescript': 'tslint', 'javascript': 'eslint'}
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'never'
let g:ale_linter_aliases = {}
let g:ale_linters = {'typescript': ['tslint'], 'javascript': ['eslint']}
let g:ale_open_list = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_offset = 1000000
let g:ale_sign_warning = '--'
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']
let g:ale_warn_about_trailing_whitespace = 1
`

**Comment by w0rp** [comment]
@yudisupriyadi123 Cool, I'll check out that repository later.

@ryanbas21 I'll need more details. Your output above doesn't tell me much of anything. You say you're getting a message? From where?

**Comment by ryanbas21** [comment]
@w0rp 
I have lint on save enabled, and i am in a .tsx file. When I try to save the file, I get the message `unknown function: tslint`

I currently have `let g:ale_fixers = { 'javascript' : 'eslint', 'typescript' : 'tslint' } `. If i set it to use `prettier` it works, but tslint fails. so `let g:ale_fixers = { 'javascript' : 'eslint', 'typescript' : 'prettier' } ` works, but at work we don't use prettier unfortunately.

Not sure why this is the issue though.

**Comment by w0rp** [comment]
Where do you get that message? Do you get an error? Is it echoed to the console? Is it in the quickfix window?

**Comment by ryanbas21** [comment]
Sorry, yes I get it in the console, [bottom of the screen](https://imgur.com/a/xMKWX), guess it would be called the quickfix screen.

**Comment by w0rp** [comment]
Ah, I see. Your problem is that you configured `'tslint'` as as fixer, and there's no function with that name. Don't do that. There isn't a `tslint` fixer.

**Comment by w0rp** [comment]
Open a issue to request adding `tslint --fix`. That isn't supported at the moment.

**Comment by ryanbas21** [comment]
Awesome, thanks!

**Comment by w0rp** [comment]
I'll open that issue now.

**Comment by w0rp** [comment]
Oh, you already did. Nevermind then.

**Comment by w0rp** [comment]
The remaining issue here will likely be solved by #925.
